docs/src/main/tut/generate-sources-from-idl/avro.md

```diff
@@ -0,0 +1,181 @@
+---
+layout: docs
+title: Generating sources from Avro
+permalink: /generate-sources-from-avro
+---
+
+# Generating sources from Avro
+
+In this section we are going to explain how we can generate the different Scala structures using the `Avro` IDL.
+
+## Protocols
+
+Let's start from the beginning, everything on `Avro` should be declared inside a `protocol`. 
+
+The name of that protocol will be the name of our Scala file.
+
+```avroidl 
+protocol People { 
+ ... 
+} 
+```
+
+***srcGen =>***
+
+`People.scala`
+
+Furthermore, the `protocol` can have a `namespace` which will be our Scala package:
+
+```avroidl
+@namespace("example.protocol")
+protocol People {
+ ...
+}
+```
+ 
+***srcGen =>***
+
+`example.protocol.People.scala`
+
+## Messages
+
+On `Avro`, the messages are declared with the keyword `record` and contains different fields inside. 
+The `record` will be translated to a `case class` with the same fields on it:
+
+```avroidl
+record Person {
+  string name;
+  int age;
+  boolean crossfitter;
+}
+```
+
+***srcGen =>***
+
+```scala
+final case class Person(name: String, age: Int, crossfitter: Boolean)
+```
+
+## Enums
+
+`Avro` supports `enum`s too and they are translated to a Scala `Enumeration`:
+
+```avroidl
+enum Errors {
+  NotFound, Duplicated, None
+}
+```
+
+***srcGen =>***
+
+```scala
+final object Errors extends Enumeration {
+  type Errors = Value
+  val NotFound, Duplicated, None = Value
+}
+```
+
+## Unions
+
+`Unions` are a complex `Avro` type for fields inside `record`s. 
+As its name suggest, it represents a type composed by another types.
+
+Depending on the types composing the `union`, `Mu` will interpret it on different ways:
+
+### Optional fields
+
+When we add a **`null`** to a `union` expression, we'll get a Scala `Option` of the other types declared along the `null`:
+
+```avroidl
+record PeopleRequest {
+  union {null, string} name;
+}
+```
+
+***srcGen =>***
+
+```scala
+final case class PeopleRequest(name: Option[String])
+```
+
+### Eithers
+
+When we join **`two non-null types`** on a `union` we'll get an Scala `Either` with the same types order:
+
+```avroidl
+record PeopleResponse {
+  union { Errors, Person } result;
+}
+```
+
+***srcGen =>***
+  
+```scala
+final case class PeopleResponse(result: Either[Errors.Value, Person])
+```
+
+### Coproducts
+
+And finally, when we have **`three or more non-null types`** on a single `union`, 
+we'll have a [shapeless](https://github.com/milessabin/shapeless/wiki/Feature-overview:-shapeless-2.0.0)' `Coproduct` on the same order as well:
+
+```avroidl
+record PeopleResponse {
+  union{ string, int, Errors } result;
+}
+```
+
+***srcGen =>***
+
+```scala
```

Comment:
Could we make these blocks as `tut` blocks so we will be forced to keep up them up-to-date?